File: course/string/V.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = int64_t;
using ld = long double;
using ii = pair<int, int>;

const char SPACE = ' ', ENDL = '\n';

//////////////////////////////////////////////////////////////////

template <typename T>
T read() {
    T value;
    cin >> value;
    return value;
}

int ri() {
    return read<int>();
}

ll rll() {
    return read<ll>();
}

string rs() {
    return read<string>();
}

template<typename T>
vector<T> rv(int size) {
    vector<T> a(size);
    for (int i = 0; i < size; ++i) {
        a[i] = read<T>();
    }
    return a;
}

//////////////////////////////////////////////////////////////////

bool yn(bool result, string yes = "YES", string no = "NO") {
	cout << (result ? yes : no) << ENDL;
	return result;
}

//////////////////////////////////////////////////////////////////

vector<string> split(string& s, string seps) {
    vector<string> parts;

    string part = "";
    auto add_part = [&]() {
        parts.push_back(part);
        part = "";
    };

    for (char ch : s) {
        if (string::npos != seps.find(ch)) {
            add_part();
        } else {
            part += ch;
        }
    }

    add_part();

    return parts;
}

//////////////////////////////////////////////////////////////////

ld sqrt_ld(ld value) {
    return sqrt(value);
}

//////////////////////////////////////////////////////////////////

void init_io() {
    string file_name = "";
    if (!file_name.empty()) {
        string input_file_name = file_name + ".in";
        freopen(input_file_name.c_str(), "rt", stdin);

        string output_file_name = file_name + ".out";
        freopen(output_file_name.c_str(), "wt", stdout);
    } else {
        #ifndef ONLINE_JUDGE
        string input_file_name = "../../input.txt"; // reading from file in parent of parent of directory with file
        freopen(input_file_name.c_str(), "rt", stdin);
        #endif
    }

	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	cout << setprecision(10) << fixed;
}

pair<bool, vector<bool>> no_answer(false, vector<bool>(0));

pair<bool, vector<bool>> get_ans(string& prefix, string& suffix, vector<string>& words, vector<ii>& len_to_indexes) {
    int total_len = prefix.length() + 1;

    string pref_common = prefix.substr(1, total_len - 2);
    string suff_common = suffix.substr(0, total_len - 2);

    if (pref_common != suff_common) {
        return no_answer;
    }

    string total = prefix + suffix.back();

    vector<bool> is_suffixes(words.size());
    for (int len = 1; len < total_len; ++len) {
        int first = len_to_indexes[len].first, second = len_to_indexes[len].second;
        string& first_word = words[first];
        string& second_word = words[second];

        string total_pref = total.substr(0, len);
        string total_suff = total.substr(total_len - len, len);

        bool first_pref = (first_word == total_pref);
        bool first_suff = (first_word == total_suff);

        bool second_pref = (second_word == total_pref);
        bool second_suff = (second_word == total_suff);

        if (first_pref && second_suff) {
            is_suffixes[first] = false;
            is_suffixes[second] = true;
        } else if (first_suff && second_pref) {
            is_suffixes[first] = true;
            is_suffixes[second] = false;
        } else {
            return no_answer;
        }
    }

    return { true, is_suffixes };
}

vector<bool> get_is_suffixes(vector<string>& words) {
    int length = words.size() / 2 + 1;

    vector<ii> len_to_indexes(length, { -1, -1 });
    for (int i = 0; i < words.size(); ++i) {
        string& word = words[i];
        int len = word.length();

        ii& indexes = len_to_indexes[len];
        if (-1 == indexes.first) indexes.first = i;
        else indexes.second = i;
    }

    string& max1 = words[len_to_indexes[length - 1].first];
    string& max2 = words[len_to_indexes[length - 1].second];

    auto ans12 = get_ans(max1, max2, words, len_to_indexes);
    auto ans21 = get_ans(max2, max1, words, len_to_indexes);

    return ans12.first ? ans12.second : ans21.second;
}

void solve() {
    int n = ri();
    int size = 2 * n - 2;
    vector<string> words = rv<string>(size);

    vector<bool> is_suffixes = get_is_suffixes(words);
    for (bool is_suffix : is_suffixes) {
        cout << (is_suffix ? 'S' : 'P');
    }
    cout << ENDL;
}

int main() {
    init_io();
    solve();
    return 0;
}

File: course/string/V.java
import java.awt.Point;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.Locale;
import java.util.StringTokenizer;

public class V {

    private void solve() {
        int n = readInt();
        int size = 2 * n - 2;
        String[] s = new String[size];
        for (int i = 0; i < size; ++i) {
            s[i] = readString();
        }

        boolean[] isSuffixes = getIsSuffixes(s);
        for (boolean isSuffix : isSuffixes) {
            out.print(isSuffix ? 'S' : 'P');
        }
    }

    boolean[] getIsSuffixes(String[] words) {
        int length = words.length / 2 + 1;

        Point[] lenToIndexes = new Point[length];
        for (int len = 1; len < length; ++len) {
            lenToIndexes[len] = new Point(-1, -1);
        }

        for (int i = 0; i < words.length; ++i) {
            int len = words[i].length();

            Point indexes = lenToIndexes[len];
            if (-1 == indexes.x) indexes.x = i;
            else indexes.y = i;
        }

        String max1 = words[lenToIndexes[length - 1].x];
        String max2 = words[lenToIndexes[length - 1].y];

        boolean[] isSuffixes = getIsSuffixes(max1, max2, words, lenToIndexes);
        if (null == isSuffixes) {
            isSuffixes = getIsSuffixes(max2, max1, words, lenToIndexes);
        }

        return isSuffixes;
    }

    boolean[] getIsSuffixes(String prefix, String suffix, String[] words, Point[] lenToIndexes){
        String prefixCommon = prefix.substring(1);
        String suffixCommon = suffix.substring(0, suffix.length() - 1);

        if (!prefixCommon.equals(suffixCommon)) return null;

        String total = prefix.charAt(0) + suffix;

        boolean[] isSuffixes = new boolean[words.length];
        for (int len = 1; len < total.length(); ++len) {
            int first = lenToIndexes[len].x, second = lenToIndexes[len].y;
            String firstWord = words[first], secondWord = words[second];

            boolean firstPrefix = total.startsWith(firstWord), firstSuffix = total.endsWith(firstWord);
            boolean secondPrefix = total.startsWith(secondWord), secondSuffix = total.endsWith(secondWord);

            if (firstPrefix && secondSuffix) {
                isSuffixes[first] = false;
                isSuffixes[second] = true;
            } else if (firstSuffix && secondPrefix) {
                isSuffixes[first] = true;
                isSuffixes[second] = false;
            } else {
                return null;
            }
        }

        return isSuffixes;
    }

    //////////////////////////////////////////////////////////////////

    long sqrtLong(long x) {
        long root = (long)Math.sqrt(x);
        while (root * root > x) --root;
        while ((root + 1) * (root + 1) <= x) ++root;
        return root;
    }

    //////////////////////////////////////////////////////////////////

    private boolean yesNo(boolean yes) {
        return yesNo(yes, "YES", "NO");
    }

    private boolean yesNo(boolean yes, String yesString, String noString) {
        out.println(yes ? yesString : noString);
        return yes;
    }

    //////////////////////////////////////////////////////////////////

    private long readLong() {
        return Long.parseLong(readString());
    }

    private int[] readIntArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; ++i) a[i] = readInt();
        return a;
    }

    private int readInt() {
        return Integer.parseInt(readString());
    }

    private String readString() {
        while (!tok.hasMoreTokens()) {
            String nextLine = readLine();
            if (null == nextLine) return null;
            tok = new StringTokenizer(nextLine);
        }

        return tok.nextToken();
    }

    private String readLine() {
        try {
            return in.readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    //////////////////////////////////////////////////////////////////

    private BufferedReader in;
    private StringTokenizer tok;
    private PrintWriter out;

    private void initFileIO(String inputFileName, String outputFileName) throws FileNotFoundException {
        in = new BufferedReader(new FileReader(inputFileName));
        out = new PrintWriter(outputFileName);
    }

    private void initConsoleIO() throws IOException {
        in = new BufferedReader(new InputStreamReader(System.in));
        out = new PrintWriter(System.out);
    }

    private void initIO() throws IOException {
        Locale.setDefault(Locale.US);

        String fileName = "";

        if (!fileName.isEmpty()) {
            initFileIO(fileName + ".in", fileName + ".out");
        } else {
            if (new File("input.txt").exists()) {
                initFileIO("input.txt", "output.txt");
            } else {
                initConsoleIO();
            }
        }

        tok = new StringTokenizer("");
    }

    //////////////////////////////////////////////////////////////////

    private void run() {
        try {
            long timeStart = System.currentTimeMillis();

            initIO();
            solve();
            out.close();

            long timeEnd = System.currentTimeMillis();
            System.err.println("Time(ms) = " + (timeEnd - timeStart));
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    public static void main(String[] args) {
        new V().run();
    }
}
